src/System.Security.Cryptography.Csp/src/Internal/Cryptography/BasicSymmetricCipherCsp.cs

```diff
@@ -0,0 +1,120 @@
+
+            byte[] currentIv = cipherMode.GetCipherIv(iv);
+            if (currentIv != null)
+            {
+                SetKeyParameter(_hKey, CryptGetKeyParamQueryType.KP_IV, currentIv);
+            }
+
+            if (effectiveKeyLength != 0)
+            {
+                SetKeyParameter(_hKey, CryptGetKeyParamQueryType.KP_EFFECTIVE_KEYLEN, effectiveKeyLength);
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                SafeKeyHandle hKey = _hKey;
+                _hKey = null;
+                if (hKey != null)
+                {
+                    hKey.Dispose();
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        public override int Transform(byte[] input, int inputOffset, int count, byte[] output, int outputOffset)
+        {
+            return Transform(input, inputOffset, count, output, outputOffset, false);
+        }
+
+        public override byte[] TransformFinal(byte[] input, int inputOffset, int count)
+        {
+            Debug.Assert(input != null);
+            Debug.Assert(inputOffset >= 0);
+            Debug.Assert(count >= 0);
+            Debug.Assert((count % BlockSizeInBytes) == 0);
+            Debug.Assert(input.Length - inputOffset >= count);
+
+            byte[] output = new byte[count];
+            if (count != 0)
+            {
+                int numBytesWritten = Transform(input, inputOffset, count, output, 0, true);
+                Debug.Assert(numBytesWritten == count);  // Our implementation of Transform() guarantees this.
+            }
+
+            return output;
+        }
+
+        private int Transform(byte[] input, int inputOffset, int count, byte[] output, int outputOffset, bool isFinal)
+        {
+            Debug.Assert(input != null);
+            Debug.Assert(inputOffset >= 0);
+            Debug.Assert(count > 0);
+            Debug.Assert((count % BlockSizeInBytes) == 0);
+            Debug.Assert(input.Length - inputOffset >= count);
+            Debug.Assert(output != null);
+            Debug.Assert(outputOffset >= 0);
+            Debug.Assert(output.Length - outputOffset >= count);
+
+            int numBytesWritten;
+            if (_encrypting)
+            {
+                numBytesWritten = EncryptData(_hKey, input, inputOffset, count, output, outputOffset, output.Length - outputOffset, isFinal);
+            }
+            else
+            {
+                numBytesWritten = DecryptData(_hKey, input, inputOffset, count, output, outputOffset, output.Length - outputOffset);
+            }
+
+            return numBytesWritten;
+        }
+
+        private static SafeKeyHandle ImportCspBlob(int algId, byte[] rawKey, bool addNoSaltFlag)
+        {
+            SafeKeyHandle safeKeyHandle;
+
+            SafeProvHandle safeProvHandleTemp = AcquireSafeProviderHandle();
```

Comment:
Good catch. I created a field for the provhandle (as not sure of lifetime semantics between provider and keys) and clean up in Dispose()